--- scikit-learn-0.19.1/sklearn/metrics/cluster/setup.py~	2017-10-21 23:28:04.000000000 +0000
+++ scikit-learn-0.19.1/sklearn/metrics/cluster/setup.py	2018-06-14 15:11:12.489110121 +0000
@@ -5,7 +5,7 @@
 
 
 def configuration(parent_package="", top_path=None):
-    config = Configuration("metrics/cluster", parent_package, top_path)
+    config = Configuration("metrics.cluster", parent_package, top_path)
     libraries = []
     if os.name == 'posix':
         libraries.append('m')
